src/compiler/utilities.ts

```diff
@@ -2851,13 +2851,26 @@ namespace ts {
         let lineCount: number;
         let linePos: number;
 
+        function updateLineCountAndPosFor(s: string) {
+            const lineStartsOfS = computeLineStarts(s);
+            if (lineStartsOfS.length > 1) {
+                lineCount = lineCount + lineStartsOfS.length - 1;
+                linePos = output.length - s.length + last(lineStartsOfS);
+                lineStart = (linePos - output.length) === 0;
+            }
+            else {
+                lineStart = false;
+            }
+        }
+
         function write(s: string) {
             if (s && s.length) {
                 if (lineStart) {
-                    output += getIndentString(indent);
+                    s = getIndentString(indent) + s;
                     lineStart = false;
                 }
                 output += s;
+                updateLineCountAndPosFor(s);

```

Comment:
So we are walking now every string we write in the emitter to find lineStarts? why is this needed? do you know the perf impact on the emitter?

Reply:
If we don't count the lines and the length of the last one, we don't set the line and column correctly for get line and column, resulting in incorrect sourcemaps.

Reply:
i understand, but how often are these functions called with multi-line text? can we split these to have their own writer function to avoid doing this on every token we write?

Reply:
Probably best to discuss this in #24704